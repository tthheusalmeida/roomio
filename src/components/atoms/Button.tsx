"use client";

import { mergeClassNames } from "@/utils/classNames";
import { VscLoading } from "react-icons/vsc";

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  children?: React.ReactNode;
  variant?: "primary" | "secondary";
  size?: "small" | "normal";
  shadow?: boolean;
  isFull?: boolean;
  isLoading?: boolean;
}

const buttonVariants = {
  primary:
    "bg-violet-900 text-[var(--color-foreground)] hover:enabled:bg-violet-800 disabled:bg-violet-400 disabled:text-violet-50",
  secondary:
    "bg-violet-100 text-violet-800 hover:enabled:bg-violet-200 disabled:bg-white disabled:text-violet-100",
};

const isLodingVariants = {
  primary: "animate-spin text-white",
  secondary: "animate-spin text-violet-400",
};

const sizeVariants = {
  small: "px-2 py-1 gap-2",
  normal: "px-4 py-2 gap-4",
};

export default function Button({
  children,
  className,
  variant,
  shadow = false,
  isFull = true,
  isLoading = false,
  size = "normal",
  ...rest
}: ButtonProps) {
  const componentClasses = [
    "flex relative inline-flex justify-center items-center",
    sizeVariants[size],
    "text-sm font-medium",
    "transition-all duration-200 cursor-pointer",
    "disabled:cursor-not-allowed",
    "rounded-sm",
    isFull ? "w-full" : "",
    shadow ? "shadow-md" : "",
    variant ? buttonVariants[variant] : "",
    className ?? "",
  ].join(" ");

  const childrenClasses = [
    "flex items-center gap-4",
    isLoading ? "invisible" : "visible",
  ].join(" ");

  const loadingClasses = [
    "absolute",
    "left-1/2 -translate-x-1/2",
    variant ? isLodingVariants[variant] : "",
  ].join(" ");

  return (
    <button className={mergeClassNames(componentClasses)} {...rest}>
      <span className={childrenClasses}>{children}</span>
      {isLoading && <VscLoading size={24} className={loadingClasses} />}
    </button>
  );
}
