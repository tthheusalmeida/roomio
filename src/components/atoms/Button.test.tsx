import { render, screen, fireEvent, act } from "@testing-library/react";
import Button from "./Button";
import { VscLoading } from "react-icons/vsc";

// Mock do Ã­cone para facilitar testes
jest.mock("react-icons/vsc", () => ({
  VscLoading: (props: any) => <span data-testid="loading-icon" {...props} />,
}));

describe("Button component", () => {
  it("renders with children text", () => {
    const { asFragment } = render(<Button>Click me</Button>);
    expect(screen.getByText("Click me")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("applies variant, size, shadow and isFull classes", () => {
    render(
      <Button variant="primary" size="small" shadow isFull={false}>
        Test
      </Button>
    );
    const btn = screen.getByText("Test").closest("button");
    expect(btn).toHaveClass("bg-violet-900"); // primary variant
    expect(btn).toHaveClass("px-2 py-1"); // small size
    expect(btn).toHaveClass("shadow-md"); // shadow
    expect(btn).not.toHaveClass("w-full"); // isFull false
  });

  it("shows loading state correctly", () => {
    render(
      <Button variant="secondary" isLoading>
        Loading
      </Button>
    );

    const children = screen.getByText("Loading");
    expect(children).toHaveClass("invisible");

    expect(screen.getByTestId("loading-icon")).toBeInTheDocument();
  });

  it("calls onClick and triggers pulsing animation", () => {
    jest.useFakeTimers();
    const handleClick = jest.fn();
    render(<Button onClick={handleClick}>Click</Button>);

    const btn = screen.getByText("Click").closest("button")!;

    act(() => {
      fireEvent.click(btn);
    });

    expect(handleClick).toHaveBeenCalled();
    expect(btn).toHaveClass("pulse-once");

    act(() => {
      jest.advanceTimersByTime(200);
    });

    expect(btn).not.toHaveClass("pulse-once");

    jest.useRealTimers();
  });
});
