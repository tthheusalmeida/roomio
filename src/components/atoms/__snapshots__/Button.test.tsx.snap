// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Button component renders with children text 1`] = `
<DocumentFragment>
  <button
    class="relative inline-flex justify-center items-center px-4 py-2 gap-4 text-sm font-medium transition-all duration-200 cursor-pointer disabled:cursor-not-allowed rounded-sm w-full"
  >
    <span
      class="flex items-center gap-4 visible"
    >
      Click me
    </span>
  </button>
</DocumentFragment>
`;
